curvefs/src/client/rpcclient/metacache.cpp

```diff
@@ -183,6 +189,176 @@ bool MetaCache::IsLeaderMayChange(const CopysetGroupID &groupID) {
     return iter->second.LeaderMayChange();
 }
 
+void MetaCache::UpdateCopysetInfo(const CopysetGroupID &groupID,
+                                  const CopysetInfo<MetaserverID> &csinfo) {
+    const auto key = CalcLogicPoolCopysetID(groupID);
+
+    WriteLockGuard wl(rwlock4copysetInfoMap_);
+    copysetInfoMap_[key] = csinfo;
+}
+
+bool MetaCache::ListPartitions(uint32_t fsID) {
+    if (init_) {
+        return true;
+    }
+
+    WriteLockGuard wl4PartitionMap(rwlock4Partitions_);
+    WriteLockGuard wl4CopysetMap(rwlock4copysetInfoMap_);
+
+    if (init_) {
+        return true;
+    } else {
+        fsID_ = fsID;
+        LOG(INFO) << "init partition and copyset infos for {fsid:" << fsID
+                  << "}";
+    }
+
+    PatitionInfoList partitionInfos;
+    std::map<PoolIDCopysetID, CopysetInfo<MetaserverID>> copysetMap;
+    if (!DoListOrCreatePartitions(true, &partitionInfos, &copysetMap)) {
+        return false;
+    }
+
+    DoAddPartitionAndCopyset(partitionInfos, copysetMap);
+
+    init_ = true;
+
+    LOG(INFO) << "init partition and copyset infos for {fsid:" << fsID
+              << "} ok";
+    return true;
+}
+
+bool MetaCache::CreatePartitions(int currentNum,
+                                 PatitionInfoList *newPartitions) {
+    std::lock_guard<Mutex> lg(createMutex_);
+
+    // already create
+    {
+        ReadLockGuard rl(rwlock4Partitions_);
+        if (partitionInfos_.size() > currentNum) {
+            return true;
+        }
+    }
+
+    // create partition
+    PatitionInfoList partitionInfos;
+    std::map<PoolIDCopysetID, CopysetInfo<MetaserverID>> copysetMap;
+    if (!DoListOrCreatePartitions(false, &partitionInfos, &copysetMap)) {
+        return false;
+    }
+
+    // add partition and copyset info
+    WriteLockGuard wl4PartitionMap(rwlock4Partitions_);
+    WriteLockGuard wl4CopysetMap(rwlock4copysetInfoMap_);
+    DoAddPartitionAndCopyset(partitionInfos, copysetMap);
+}
+
+bool MetaCache::DoListOrCreatePartitions(
```

Comment:
so many RPC interactive, can we just send one RPC and get what we want?

Reply:
yes, need to be modify. I mark todo here